torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -0,0 +1,191 @@
+// updateOutput, updateGradInput Kernels ported from Sergey Zagoruyko's pyinn, which itself was a
+// port from Caffe
+
+#include "THCUNN.h"
+#include "THCDeviceTensor.cuh"
+#include "THCDeviceTensorUtils.cuh"
+#include "THCNumerics.cuh"
+#include "THCReduceApplyUtils.cuh"
+#include "THCSortUtils.cuh"
+#include "THCTensorMathReduce.cuh"
+#include "SharedMem.cuh"
+#include "common.h"
+
+template <typename T, typename IndexType>
+__global__ void spatialDepthwiseConvolutionUpdateOutput(
+    const THCDeviceTensor<T, 4> input,
+    THCDeviceTensor<T, 4> output,
+    const THCDeviceTensor<T, 4> weight,
+    const THCDeviceTensor<T, 1> bias,
+    bool biasEnabled,
+    IndexType totalElements,
+    const int outputChannels,
+    const int depthwiseMultiplier,
+    const int inputWidth, const int inputHeight,
+    const int outputWidth, const int outputHeight,
+    const int kernelWidth, const int kernelHeight,
+    const int strideWidth, const int strideHeight,
+    const int padWidth, const int padHeight,
+    const int dilationWidth, const int dilationHeight)
+{
+  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < totalElements;
+       linearIndex += gridDim.x * blockDim.x) {
+
+    const int n = linearIndex / outputChannels / outputHeight / outputWidth;
+    const int c = (linearIndex / outputHeight / outputWidth) % outputChannels;
+    const int h = (linearIndex / outputWidth) % outputHeight;
+    const int w = linearIndex % outputWidth;
+
+    const int inputChannel = c / depthwiseMultiplier;
+    const int inputChannels = outputChannels / depthwiseMultiplier;
+
+    int weightOffset = c * kernelHeight * kernelWidth;
+
+    T value = biasEnabled ? bias.data()[c] : ScalarConvert<int, T>::to(0);
+    for (int kH = 0; kH < kernelHeight; ++kH) {
+      for (int kW = 0; kW < kernelWidth; ++kW) {
+        const int h_in = -padHeight + h * strideHeight + kH * dilationHeight;
+        const int w_in = -padWidth + w * strideWidth + kW * dilationWidth;
+
+        if ((h_in >= 0) && (h_in < inputHeight) && (w_in >= 0) && (w_in < inputWidth)) {
+          const IndexType offset = ((n * inputChannels + inputChannel) * inputHeight + h_in) *
+                                    inputWidth + w_in;
+          value = THCNumerics<T>::add(
+            value,
```

Comment:
for halfs, accumulation here should really be in float, otherwise it is truncated to half after each multiply-add. Accumulation in float should not cost anything in terms of performance, and should provide better accuracy (as well as protect against intermediate overflows/underflows).

Reply:
Will add an accreal parameter.

Reply:
Actually, @ngimel --> it looks like we have `accreal=double` for `float` types, do we want to do that as well? or only accumulate half in float?

Reply:
My recollection is ```accreal=double``` for ```float``` in THNN/TH, and ```accreal=float``` for ```float``` in THCUNN/THC, but I might be mistaken. We definitely don't want to accumulate in double for float, double performance is bad on many cards, but we do want float accumulation for halfs.

Reply:
@ngimel ah yes, was looking at the CPU file accidentally instead :(